cmd/admin-revoker/main.go

```diff
@@ -211,7 +237,117 @@ func (r *revoker) revokeByReg(ctx context.Context, regID int64, reasonCode revoc
 }
 
 func (r *revoker) revokeMalformedBySerial(ctx context.Context, serial string, reasonCode revocation.Reason) error {
-	return r.revokeCertificate(ctx, core.Certificate{Serial: serial}, reasonCode)
+	return r.revokeCertificate(ctx, core.Certificate{Serial: serial}, reasonCode, false)
+}
+
+// blockByPrivateKey blocks future issuance for certificates with a a public key
+// matching the SubjectPublicKeyInfo hash generated from the PublicKey embedded
+// in privateKey. The embedded PublicKey will be verfied as an actual match for
+// the provided private key before any blocking takes place. This method does
+// not revoke any certificates directly. However, 'bad-key-revoker', which
+// references the 'blockedKeys' table, will eventually revoke certificates with
+// a matching SPKI hash.
+func (r *revoker) blockByPrivateKey(ctx context.Context, privateKey crypto.Signer) error {
+	err := privatekey.Verify(privateKey)
+	if err != nil {
+		return err
+	}
+
+	spkiHash, err := getPublicKeySPKIHash(privateKey.Public())
+	if err != nil {
+		return err
+	}
+
+	u, err := user.Current()
+	if err != nil {
+		return err
+	}
+
+	req := &sapb.AddBlockedKeyRequest{
+		KeyHash:   spkiHash,
+		Added:     r.clk.Now().UnixNano(),
+		Source:    "admin-revoker",
+		Comment:   fmt.Sprintf("blocked by %s", u),
+		RevokedBy: 0,
+	}
+
+	_, err = r.sac.AddBlockedKey(ctx, req)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// revokeByPrivateKey revokes all certificates with a public key matching the
+// SubjectPublicKeyInfo hash generated from the PublicKey embedded in
+// privateKey. The embedded PublicKey will be verfied as an actual match for the
+// provided private key before any revocation takes place. The provided key will
+// not be added to the 'blockedKeys' table, this is done to avoid a race between
+// 'admin-revoker' and 'bad-key-revoker'. You MUST call blockByPrivateKey after
+// calling this function, on pain of violating the BRs.
+func (r *revoker) revokeByPrivateKey(ctx context.Context, privateKey crypto.Signer) error {
+	err := privatekey.Verify(privateKey)
+	if err != nil {
+		return err
+	}
+
+	spkiHash, err := getPublicKeySPKIHash(privateKey.Public())
+	if err != nil {
+		return err
+	}
```

Comment:
Rather than relying on `privatekey.Verify` previously having been called, it would be nice if `privatekey.Verify` returns the public key upon successful verification.

Actually, exploring that idea of "use the type system to avoid mistakes": What if `privatekey.Load()` performs a verification on every load, and returns an error if verification fails? Then we can shrink the public API of the `privatekey` package, and also be sure we never get private keys where the public key hasn't been verified to match.

Reply:
Good idea, I went ahead and implemented it.